File: 92-reverse-linked-list-ii/92-reverse-linked-list-ii.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverse(ListNode curr){
        ListNode prev = null, nxt = null;
        
        while(curr != null){
            nxt = curr.next;
            curr.next = prev;
            
            prev = curr;
            curr = nxt;
        }
        
        return prev;
    }
    
    public ListNode reverseBetween(ListNode head, int left, int right) {
        
        if (head == null || head.next == null) return head;
        
        ListNode curr = head, prev = null;
        int count = 1;
        
        while(count != left){
            prev = curr;
            curr = curr.next;
            count += 1;
        }
        
        ListNode start = curr;
        while(count != right){
            curr = curr.next;
            count += 1;
        }
        
        ListNode tail = curr.next;
        curr.next = null;
        
        ListNode newhead = reverse(start);
        
        if(prev != null) prev.next = newhead;
        
        curr = newhead;
        while(curr.next != null){
            curr = curr.next;
        }
        
        curr.next = tail;
        
        if(left == 1) return newhead;
        return head;
    }
}

File: 92-reverse-linked-list-ii/92-reverse-linked-list-ii.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseBetween(self, head: Optional[ListNode], left: int, right: int) -> Optional[ListNode]:
        
        def reverse(curr):
            prev, nxt = None, None
            while curr:
                nxt = curr.next
                curr.next = prev
                
                prev = curr
                curr = nxt
            return prev
        
        
        curr, prev = head, None
        count = 1
        
        while count != left:
            prev = curr
            curr = curr.next
            count += 1
        
        start = curr
        while count != right:
            curr = curr.next
            count += 1
            
        tail = curr.next
        curr.next = None
        
        newhead = reverse(start)
        
        if prev:
            prev.next = newhead
        
        curr = newhead
        while curr.next:
            curr = curr.next
        
        curr.next = tail
        
        if left==1:
            return newhead
        return head
    